Automattic/woocommerce-payments
# Initial checkout block support

Fixes #652

### Changes proposed in this Pull Request

This PR was supposed to be a spike, related to Stripe's React elements, but grew into a PR with support for WooCommerce Blocks. 😲 

This was done by following https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/main/docs/extensibility/payment-method-integration.md for the checkout block, and isolating some functionality from the existing `wcpay-checkout.js`.

### Testing instructions

1. Make sure that the classic checkout is not broken. This includes:
    - Testing with both a simple card (`**** 4242`) and an SCA-enabled card (ex. `4000002760003184`).
    - Going through a successful workflow.
    - Checking whether all error messages (or failure scenarios) are handled in the correct location on the page (you'll know it ;) ).
2. Install https://github.com/woocommerce/woocommerce-gutenberg-products-block and create a page with it.
3. Retry all of the tests from the classic checkout within the checkout block.

### Notes and questions

#### All scripts and dependencies are loaded within Gutenberg

Is this expected, or it should somehow happen dynamically? Even when composing a new post without a checkout block, Stripe's API and all necessary scripts are loaded.

#### Erroneous state after confirmation

<img width="797" alt="error-notice" src="https://user-images.githubusercontent.com/5311119/84786093-f7a6b480-aff4-11ea-8f5b-8f6b18b65865.png">

If there is an error during 3DS authentication (through the `onCheckoutAfterProcessingWithSuccess` handler):

1. The error message looks broken. The X button is misaligned.
2. The page does not scroll to the error message, so it remains above the fold.
3. Even though the confirmation failed, the submit button displays a seemingly successful check-mark.<br /><img width="479" alt="fields-after-error" src="https://user-images.githubusercontent.com/5311119/84786387-5cfaa580-aff5-11ea-9821-8f27650dfb68.png">

#### Webpack/Babel issues between checkout method

Everything, related to the new checkout (with blocks) gets compiled properly. However, trying to use `async` functions for the classic checkout fails.

Any idea why this might be happening? Both `client/checkout/blocks/index.js` and `client/checkout/classic/index.js` are loaded the same way in webpack's config.

P.S. There is a failing PHPUnit test, which I will look into, but is not blocking the review of the PR.
